src/parser.py

```diff
@@ -21,7 +32,7 @@ def __init__(
     def parse(self, program):
         program = self.preprocessor.process(program)
         tree = self.grammar.parse(program)
-        return self.visitor.visit(tree)
+        return Ast(self.visitor.visit(tree))
```

Comment:
I love the idea of adding as much typing and object orientation as possible. I would add `program: str` to the method's input so that it is fully typed, both in its input and output.